### PR TITLE
Feature/farm 365 plot repository

### PR DESCRIPTION
**Pull request by LeeMatthewHiggins**
[FARM-365 - Plot repository]

#### 📲 What

The repository interface and mock repo
Plus the minimum needed to demo the interface and Stage logic

#### 🤔 Why
		
To implement and test business logic for My plot screen. (stages)
		
#### 🛠 How
		
Bloc pattern

#### ✅ Acceptance criteria

- [x] Rebased/merged with latest changes from development and re-tested

Recommended Style Guide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

#### 🕵️‍♂️ How to test

Notes for QA

#### Reviewers

@farmsmart/amido @farmsmart/wamf

### CODE REVIEW

File: lib/data/bloc/plot/PlotDetailProvider.dart [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    final beginNext = stage != plot.stages.last;
+    _repo.completeStage(plot, stage).then((plot) {
+      final currentStage = _logic.currentStage(plot.stages);
+      if (beginNext) {
+          _repo.beginStage(plot, currentStage).then((plot){
+           _plot = plot;
+        });
+      }
+      else
+      {
```
Comment (albertmiro) [inline lib/data/bloc/plot/PlotDetailProvider.dart]: Can we move this else to be in the same line?

`} else {`

File: lib/data/bloc/plot/PlotDetailProvider.dart [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+  void _revertStageAction(PlotEntity plot, NewStageEntity stage) {
+    _repo.revertStage(plot, stage).then((plot) {
+      _plot = plot;
+    });
+  }
+
+  void _remove(PlotEntity plot){
+     _repo.remove(plot).then((success) {
+      //TODO: what if this failed?
```
Comment (albertmiro) [inline lib/data/bloc/plot/PlotDetailProvider.dart:86]: That would be done in another PR?

File: lib/data/bloc/plot/PlotListProvider.dart [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  PlotListViewModel initial() {
+    if (_snapshot == null) {
+      _snapshot = _viewModel(status: LoadingStatus.LOADING, items: []);
+      _snapshot.update();
+    }
+    return _snapshot;
+  }
+
+  PlotListViewModel _viewModel({LoadingStatus  status, List<PlotListItemViewModel> items}){
+    return PlotListViewModel(title: _title, buttonTitle: Intl.message(_Strings.addCrop), status: status, items: items, update: () => _update(_controller), add: () => _add());
```
Comment (albertmiro) [inline lib/data/bloc/plot/PlotListProvider.dart:78]: Can we split that to different lines? With a comma I think that this would be formatted automatically.

```
PlotListViewModel(
    title: _title, 
    buttonTitle: Intl.message(_Strings.addCrop), 
    status: status, 
    items: items, 
    update: () => _update(_controller), 
    add: () => _add(),
);
```

File: lib/data/bloc/plot/StageBusinessLogic.dart [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+class StageBusinessLogic {
+
+  double progress(List<NewStageEntity> stages) {
+    if (isComplete(stages.last)) {
+      return  1.0;
+    }
+    final currentIndex = currentStageIndex(stages);
+    return max(0.05,currentIndex.toDouble()  / (stages.length).toDouble());
```
Comment (albertmiro) [inline lib/data/bloc/plot/StageBusinessLogic.dart]: We can move these values to constants?

`final fullProgress = 1.0;`
`final emptyProgress = 0.05;`

File: lib/data/model/NewStageEntity.dart [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import 'package:farmsmart_flutter/data/model/article_entity.dart';
+
+class NewStageEntity {
+  final String id;
+  final ArticleEntity article;
+  final DateTime started;
+  final DateTime ended;
+
+  NewStageEntity({String id, ArticleEntity article, DateTime started, DateTime ended}) : this.id = id, this.article = article, this.started = started, this.ended = ended;
```
Comment (albertmiro) [inline lib/data/model/NewStageEntity.dart]: Missing comma before `}` so in this way the IDE can format that. There're some places in the PR where this happens, I haven't added comments at each one, if you want I can go through them and add the missing comma at the end.

```suggestion
  NewStageEntity({
    String id,
    ArticleEntity article,
    DateTime started,
    DateTime ended,
  })  : this.id = id,
        this.article = article,
        this.started = started,
        this.ended = ended;
```

File: lib/ui/common/ActionSheet.dart [hunk 6]
```diff
@@ -107,6 +103,12 @@ class ActionSheet extends StatefulWidget {
   final ActionSheetViewModel _viewModel;
   final ActionSheetStyle _style;
 
+  static present(ActionSheet sheet, BuildContext context) {
+    showModalBottomSheet(backgroundColor: Colors.transparent,
```
Comment (davidfranquet) [inline lib/ui/common/ActionSheet.dart:107]: Is not compiling with this param, I think there's no `backgroundColor` on `showModalBottomSheet` func

```suggestion
    showModalBottomSheet(
```

Comment (davidfranquet) [reply lib/ui/common/ActionSheet.dart:107]: image

File: lib/data/repositories/plot/implementation/MockPlotRepository.dart [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  }
+
+  @override
+  Future<PlotEntity> getSingle(String uri) {
+    for (var i = 0; i < _plots.length; i++) {
+      if(_plots[i].id == uri){
+        return Future.value(_plots[i]);
+      }
+    }
+    return Future.error(Error());
```
Comment (davidfranquet) [inline lib/data/repositories/plot/implementation/MockPlotRepository.dart]: You can simplify this block with a `singleWhere` instead of iterating the array. `singleWhere` will take the first item in the array matching the condition that you set inside the `singleWhere` also you can catch the `orElse` param if doesn't find any, in this case I think that you need to return a null so the `orElse` I think that is not needed.
```suggestion
    final plotEntity = _plots.singleWhere((plotEntity) => plotEntity.id == uri, orElse: () => null);
    return plotEntity != null ? Future.value(plotEntity) : Future.error(Error());
```

File: lib/ui/common/SectionListView.dart [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  final int offset;
+  final ListViewSection section;
+
+  _SectionPosition(this.offset, this.section);
+  int relativeOffset(int absoluteOffset) {
+      return absoluteOffset - offset;
+  }
+}
+
+class SectionedListView extends StatelessWidget implements ListViewSection {
```
Comment (davidfranquet) [inline lib/ui/common/SectionListView.dart:20]: 👍

File: lib/ui/common/SectionListView.dart [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    final sectionPositions = sections.map((section) {
+        int offset = allItemCount;
+        allItemCount += section.length();
+        return _SectionPosition(offset, section);
+    }).toList();
+    return  SectionedListView._(sectionPositions,allItemCount,shrinkWrap,physics);
+  }
+
+  _SectionPosition sectionForIndex(int index) {
+    for (var sectionPosition in _sectionPositions.reversed) {
```
Comment (davidfranquet) [inline lib/ui/common/SectionListView.dart]: Same as the previous comment you can use `firstWhere` or `singleWhere`
```suggestion
    return _sectionPositions.reversed.firstWhere((
        sectionPosition) => sectionPosition.relativeOffset(index) >= 0,
        orElse: () => null);
```

File: lib/ui/myplot/PlotList.dart [hunk 10]
```diff
@@ -57,74 +53,72 @@ class _DefaultStyle implements PlotListStyle {
...
   @override
   Widget build(BuildContext context,
-      {PlotListStyle plotStyle = const _DefaultStyle()}) {
-    return Scaffold(
-      body: StoreConnector<AppState, MyPlotViewModel>(
-          onInit: (store) => store.dispatch(FetchCropListAction()),
-          builder: (_, viewModel) => _buildBody(context, viewModel, plotStyle),
-          converter: (store) => MyPlotViewModel.fromStore(store)),
-    );
+      {PlotListStyle style = const _DefaultStyle()}) {
```
Comment (davidfranquet) [inline lib/ui/myplot/PlotList.dart]: This should be here? Or should be inside the PlotList constructor?

**Review by davidfranquet — CHANGES_REQUESTED** [review]
Can't compile 😕 

See:  https://github.com/farmsmart/farmsmart-flutter/pull/57#discussion_r306242951

File: lib/ui/common/ActionSheet.dart (hunk 6 above)
Comment (LeeMatthewHiggins) [reply lib/ui/common/ActionSheet.dart:107]: that's why the CI´s not working

I did update my flutter :(

Comment (davidfranquet) [reply lib/ui/common/ActionSheet.dart:107]: just remove this and it works :D

File: lib/data/bloc/plot/PlotDetailProvider.dart (hunk 2 above)
Comment (LeeMatthewHiggins) [reply lib/data/bloc/plot/PlotDetailProvider.dart:86]: yep else this PR gets out of control. It may do nothing, as if it fails, the user will just have to try again.

File: lib/data/bloc/plot/StageBusinessLogic.dart (hunk 4 above)
Comment (LeeMatthewHiggins) [reply lib/data/bloc/plot/StageBusinessLogic.dart]: 👍

File: lib/data/model/NewStageEntity.dart (hunk 5 above)
Comment (LeeMatthewHiggins) [reply lib/data/model/NewStageEntity.dart]: 👍

File: lib/ui/common/ActionSheet.dart (hunk 6 above)
Comment (LeeMatthewHiggins) [reply lib/ui/common/ActionSheet.dart:107]: 👍

File: lib/data/repositories/plot/implementation/MockPlotRepository.dart (hunk 7 above)
Comment (LeeMatthewHiggins) [reply lib/data/repositories/plot/implementation/MockPlotRepository.dart]: nice

Comment (LeeMatthewHiggins) [reply lib/data/repositories/plot/implementation/MockPlotRepository.dart]: 👍

File: lib/ui/common/SectionListView.dart (hunk 9 above)
Comment (LeeMatthewHiggins) [reply lib/ui/common/SectionListView.dart]: 👍

File: lib/ui/common/ActionSheet.dart (hunk 6 above)
Comment (LeeMatthewHiggins) [reply lib/ui/common/ActionSheet.dart:107]: Need it to have the rounded corners however:(

File: lib/ui/myplot/PlotList.dart (hunk 10 above)
Comment (LeeMatthewHiggins) [reply lib/ui/myplot/PlotList.dart]: yeah should be in the constructor 👍

File: lib/data/bloc/plot/PlotListProvider.dart (hunk 3 above)
Comment (LeeMatthewHiggins) [reply lib/data/bloc/plot/PlotListProvider.dart:78]: 👍

File: lib/data/bloc/plot/PlotDetailProvider.dart (hunk 1 above)
Comment (LeeMatthewHiggins) [reply lib/data/bloc/plot/PlotDetailProvider.dart]: 👍

File: lib/ui/common/ActionSheet.dart (hunk 6 above)
Comment (davidfranquet) [reply lib/ui/common/ActionSheet.dart:107]: 👍

Comment (LeeMatthewHiggins) [reply lib/ui/common/ActionSheet.dart:107]: you'll have to do a flutter update for it to build. we need that line. just updated the CI to same version I'm using.

Comment (davidfranquet) [reply lib/ui/common/ActionSheet.dart:107]: 👌  I'll update flutter then

**Review by davidfranquet — DISMISSED** [review]

**Review by albertmiro — DISMISSED** [review]

**Review by davidfranquet — APPROVED** [review]

**Review by isaacmartinezimv — APPROVED** [review]
